lib/model/model.go

```diff
@@ -2042,14 +2042,38 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 		return err
 	}
 
-	batch := newFileInfoBatch(func(fs []protocol.FileInfo) error {
+	batchFn := func(fs []protocol.FileInfo) error {
 		if err := runner.CheckHealth(); err != nil {
 			l.Debugf("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
 			return err
 		}
 		m.updateLocalsFromScanning(folder, fs)
 		return nil
-	})
+	}
+	// Resolve items which are identical with the global state.
+	if localFlags&protocol.FlagLocalReceiveOnly != 0 {
+		oldBatchFn := batchFn // can't reference batchFn directly (recursion)
+		batchFn = func(fs []protocol.FileInfo) error {
+			for i := range fs {
+				switch gf, ok := fset.GetGlobal(fs[i].Name); {
+				case !ok:
+					continue
+				case gf.IsEquivalentOptional(fs[i], false, false, protocol.FlagLocalReceiveOnly):
+					// What we have locally is equivalent to the global file.
+					fs[i].Version = fs[i].Version.Merge(gf.Version)
+					fallthrough
+				case fs[i].IsDeleted() && gf.IsReceiveOnlyChanged():
+					// Our item is deleted and the global item is our own
+					// receive only file. We can't delete file infos, so
+					// we just pretend it is a normal deleted file (nobody
+					// cares about that).
+					fs[i].LocalFlags &^= protocol.FlagLocalReceiveOnly
```

Comment:
Wait.

1. There is a normal file. It's in sync everywhere.
2. We, a receive only device, change the file. We set the receive only changed bit and update the version. This is the global item now in our view.
3. We delete the file.
4. ???

Reply:
That's what the change in global list is about: https://github.com/syncthing/syncthing/pull/5230/files#diff-69b15624219bb5333982e6cf1df4ef78L50  
Meaning in your point 2, "This is the global item now in our view." is no longer true, as it's invalid. If there is any non-invalid file, that's global before ours. This switch branch can only happen if we are the only ones having the file (otherwise `gf.IsReceiveOnlyChanged` is false).

Reply:
Right, yes, OK. And the GreaterOrEqual in the need loop should mean that it doesn't become needed anyway.

I merge conflicted you by doing a couple of renames, will see if I can merge that for you.